app/scripts/lib/multichain-api/wallet-requestPermissions.js

```diff
@@ -0,0 +1,179 @@
+    const { chainId } = getNetworkConfigurationByNetworkClientId(
+      req.networkClientId,
+    );
+
+    const scopeString = `eip155:${parseInt(chainId, 16)}`;
+
+    const ethAccounts = ethAccountsPermission.caveats[0].value;
+
+    const caipAccounts = ethAccounts.map(
+      (account) => `${scopeString}:${account}`,
+    );
+
+    const permissions = getPermissionsForOrigin(origin);
+    const caip25endowment = permissions[Caip25EndowmentPermissionName];
+    const caip25caveat = caip25endowment?.caveats.find(
+      ({ type }) => type === Caip25CaveatType,
+    );
+    if (caip25caveat) {
+      const { optionalScopes, ...caveatValue } = caip25caveat.value;
+      const optionalScope = {
+        methods: [], // TODO grant all methods
+        notifications: [], // TODO grant all notifications
+        accounts: [],
+        // caveat values are frozen and must be cloned before modified
+        // this spread comes intentionally after the properties above
+        ...optionalScopes[scopeString],
+      };
+
+      optionalScope.accounts = Array.from(
+        new Set([...optionalScope.accounts, ...caipAccounts]),
+      );
+
+      const newOptionalScopes = {
+        ...caip25caveat.value.optionalScopes,
+        [scopeString]: optionalScope,
+      };
+
+      updateCaveat(origin, Caip25EndowmentPermissionName, Caip25CaveatType, {
+        ...caveatValue,
+        optionalScopes: newOptionalScopes,
+      });
+
+      const sessionScopes = mergeScopes(
+        caip25caveat.value.requiredScopes,
+        caip25caveat.value.optionalScopes,
+      );
+
+      Object.entries(sessionScopes).forEach(([_, { accounts }]) => {
+        accounts?.forEach((account) => {
+          const {
+            address,
+            chain: { namespace },
+          } = parseCaipAccountId(account);
+
+          if (namespace === 'eip155') {
+            ethAccounts.push(address);
+          }
+        });
+      });
+
+      grantedPermissions[RestrictedMethods.eth_accounts] = {
+        ...caip25endowment,
+        parentCapability: RestrictedMethods.eth_accounts,
+        caveats: [
+          {
+            type: CaveatTypes.restrictReturnedAccounts,
+            value: Array.from(new Set(ethAccounts)),
+          },
+        ],
+      };
+    } else {
+      const caip25grantedPermissions = grantPermissions({
+        subject: { origin },
+        approvedPermissions: {
+          [Caip25EndowmentPermissionName]: {
+            caveats: [
+              {
+                type: Caip25CaveatType,
+                value: {
+                  requiredScopes: {},
+                  optionalScopes: {
+                    [scopeString]: {
+                      methods: [], // TODO grant all methods
+                      notifications: [], // TODO grant all notifications
+                      accounts: caipAccounts,
+                    },
+                  },
+                },
+              },
+            ],
+          },
+        },
+      });
```

Comment:
Do we also want to remove the restrictedMethod.eth_accounts permission from PermissionsController state since it will be in both formats as currently written?

Reply:
No, because that would mean we need to refactor all logic that relies on the eth_accounts permission from the PermissionsController. This needs to be handled at the same time we write the migration